motor/core.py

```diff
@@ -1254,38 +1280,21 @@ def each(self, callback):
         if not callable(callback):
             raise callback_type_error
 
-        self._each_got_more(callback, None)
-
-    def _each_got_more(self, callback, future):
-        if future:
+        async def _coro():
             try:
-                future.result()
-            except Exception as error:
-                callback(None, error)
-                return
-
-        while self._buffer_size() > 0:
-            doc = next(self.delegate)  # decrements self.buffer_size
-
-            # Quit if callback returns exactly False (not None). Note we
-            # don't close the cursor: user may want to resume iteration.
-            if callback(doc, None) is False:
-                return
-
-            # The callback closed this cursor?
-            if self.closed:
-                return
-
-        if self.alive and (self.cursor_id or not self.started):
-            self._framework.add_future(
-                self.get_io_loop(),
-                self._get_more(),
-                self._each_got_more, callback)
-        else:
-            # Complete
-            self._framework.call_soon(
-                self.get_io_loop(),
-                functools.partial(callback, None, None))
+                async for doc in self:
+                    # The callback closed this cursor?
+                    if self.closed:
+                        return
+                    # Quit if callback returns exactly False (not None). Note we
+                    # don't close the cursor: user may want to resume iteration.
+                    if callback(doc, None) is False:
+                        return
+                callback(None, None)
+            except Exception as exc:
```

Comment:
This is really weird design but we have to stick to it in the interest of maintaining backwards-compatibility. In 3.0 we can just raise the error from this coroutine instead of relying on the user-supplied callback to do so.